managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -2283,73 +2283,78 @@ void internalFlushPendingMarkDeletes() {
+                    // Created the ledger, now write the last position
+                    // content
+                    MarkDeleteEntry mdEntry = lastMarkDeleteEntry;
+                    persistPositionToLedger(lh, mdEntry, new VoidCallback() {
+                        @Override
+                        public void operationComplete() {
+                            if (log.isDebugEnabled()) {
+                                log.debug("[{}] Persisted position {} for cursor {}", ledger.getName(),
+                                        mdEntry.newPosition, name);
+                            }
+                            switchToNewLedger(lh, new VoidCallback() {
+                                @Override
+                                public void operationComplete() {
+                                    callback.operationComplete();
+                                }
+
+                                @Override
+                                public void operationFailed(ManagedLedgerException exception) {
+                                    // it means it failed to switch the newly created ledger so, it should be
+                                    // deleted to prevent leak
+                                    bookkeeper.asyncDeleteLedger(lh.getId(), (int rc, Object ctx) -> {
+                                        if (rc != BKException.Code.OK) {
+                                            log.warn("[{}] Failed to delete orphan ledger {}", ledger.getName(),
+                                                    lh.getId());
+                                        }
+                                    }, null);
+                                    callback.operationFailed(exception);
+                                }
+                            });
+                        }
+
+                        @Override
+                        public void operationFailed(ManagedLedgerException exception) {
+                            log.warn("[{}] Failed to persist position {} for cursor {}", ledger.getName(),
+                                    mdEntry.newPosition, name);
 
+                            ledger.mbean.startCursorLedgerDeleteOp();
+                            bookkeeper.asyncDeleteLedger(lh.getId(), new DeleteCallback() {
+                                @Override
+                                public void deleteComplete(int rc, Object ctx) {
+                                    ledger.mbean.endCursorLedgerDeleteOp();
+                                }
+                            }, null);
+                            callback.operationFailed(exception);
+                        }
+                    });
+                }));
+            }, LedgerMetadataUtils.buildAdditionalMetadataForCursor(name));
+        } catch (Throwable t) {
+            log.error("[{}] Encountered unexpected error when creating cursor ledger", name, t);
+            callback.operationFailed(new ManagedLedgerException(t));
```

Comment:
actually I don't think we need this change here. The exception is already handled in `ledger.asyncCreateLedger`

Reply:
Correct. let me pull that back